### DISCUSSION TITLE
Questions on scaling and realtime

### DISCUSSION THREAD

**Discussion by jamauro** [pocketbase/pocketbase#1673] [Q&A]
Thanks for making pocketbase. Looks really interesting and I'll be experimenting with it and sveltekit.

I'm not an expert on SQLite so forgive my potentially dumb questions. :)

Would it be crazy to use pocketbase for a multi-tenant SaaS with realtime chat? It seems like each tenant could have its own SQLite DB and as long as an individual tenant isn't massive, it would be fine. I saw your comments on this [pocketbase & svelte YouTube video](https://www.youtube.com/watch?v=gUYBFDPZ5qk). It sounded like you thought pocketbase could handle this use case with the improvements in v0.10+

Any ideas at what point write concurrency would be an issue? If I'm understanding correctly, WAL queues the writes so you wouldn't see the `database is locked` if multiple people are writing at the same time but the write might take a while to complete.

Would LiteFS or Marmot help with write concurrency?

**Comment by ganigeorgiev** [comment]
> Would it be crazy to use pocketbase for a multi-tenant SaaS with realtime chat? It seems like each tenant could have its own SQLite DB and as long as an individual tenant isn't massive, it would be fine.

A single PocketBase instance operates only with 1 data and 1 logs SQLite db file.
If by multi-tenant you mean multiple PocketBase instances, aka. separate projects on the same server, then the throughput depends on the available hardware resources.

> Any ideas at what point write concurrency would be an issue? If I'm understanding correctly, WAL queues the writes so you wouldn't see the database is locked if multiple people are writing at the same time but the write might take a while to complete.

It depends. Writes queue up, but still in case of mixed (read and write) transactions SQLITE_BUSY errors could occur. I've tried to minimize the transaction locking times but depending on the type of queries (especially delete) in heavy concurrent scenarios it could result in SQLITE_BUSY errors.

> Would LiteFS or Marmot help with write concurrency?

I don't think so (at least not out of the box). Both as far as I know replicate/sync the database to another location. PocketBase by default cannot scale horizontally and realtime events will not be triggered by just syncing the db tables. The events are triggered only on [`app.OnModel*` app level hooks](https://pocketbase.io/docs/event-hooks/#db-hooks).

---

My advice would be to create a POC for your use case and stress test it with apachebench, hey, k6, etc. and decide on your own.

Official test bed and benchmarks will be available in the near future.

**Reply by ganigeorgiev** [reply]
_Additionally, you may also want to check some recent notes about the realtime apis I left in  https://github.com/pocketbase/pocketbase/discussions/1660#discussioncomment-4748288._
